keydata_luks.go

```diff
@@ -0,0 +1,136 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"bytes"
+	"errors"
+
+	"golang.org/x/xerrors"
+
+	"github.com/snapcore/secboot/internal/luks2"
+	"github.com/snapcore/secboot/internal/luksview"
+)
+
+// LUKS2KeyDataReader provides a mechanism to read a KeyData from a LUKS2 token.
+type LUKS2KeyDataReader struct {
+	name     string
+	slot     int
+	priority int
+	*bytes.Reader
+}
+
+func (r *LUKS2KeyDataReader) ReadableName() string {
+	return r.name
+}
+
+// KeyslotID indicates the keyslot ID associated with the token from which this
+// KeyData is read.
+func (r *LUKS2KeyDataReader) KeyslotID() int {
+	return r.slot
+}
+
+// Priority indicates the priority of the keyslot associated with the token from
+// which this KeyData is read. The default priority is 0 with higher numbers
+// indicating a higher priority.
+func (r *LUKS2KeyDataReader) Priority() int {
+	return r.priority
+}
+
+// NewLUKS2KeyDataReader is used to read a LUKS2 token containing key data with
+// the specified name on the specified LUKS2 container.
+func NewLUKS2KeyDataReader(devicePath, name string) (*LUKS2KeyDataReader, error) {
```

Comment:
I never remarked this but golang style usually put constructs/New functions first after the struct definition before the methods. secboot doesn't seem to follow this convention. Might be easier to do a separate PR to address this at some point

Reply:
Thanks, I've reordered this one.